# Laboratorio 8 Boris Juárez Brolo

In [4]:
import pandas as pd
import numpy as np
from scipy import stats


In [8]:
Data_titanic = pd.read_csv('drive-download-20191016T191440Z-001/titanic_MD.csv')
missing = []

## Missing Values Report

### Array creado para evaluar los valores faltantes en las columnas del data set

In [9]:
for i in Data_titanic.columns: 
    if pd.isnull(Data_titanic[i]).any():
        missing.append(i)
print(missing)

['Age', 'SibSp', 'Parch', 'Fare', 'Embarked']


##### Los valores faltantes de la edad, Parch y Embarked se dejan vacios mientras que los valores faltantes en SibSp toman la forma de Na

## Funciones de limpieza

In [10]:
def pairwise_deletion(dataframe, columns): 
    dataframe = dataframe.dropna(subset=columns)
    return dataframe

def imputation_mode(frame, column_name=None):
    dataframe = frame
    if column_name==None:
        for i in range(183):
            dataframe = dataframe.fillna(dataframe.mode().iloc[i])
    else:
        dataframe[[column_name]] = dataframe[[column_name]].fillna(dataframe.mode().iloc[0])
    return dataframe

def outliers_standard(dataframe, column, zscore):
    dataframe = dataframe[(np.abs(stats.zscore(dataframe[[column]])) < zscore).all(axis=1)]
    return dataframe

def imputation_mean(frame, column_name=None):
    dataframe = frame
    if column_name==None:
        for i in range(7):
            dataframe = dataframe.fillna(dataframe.mean().iloc[i])
    else:
        dataframe[[column_name]] = dataframe[[column_name]].fillna(dataframe[[column_name]].mean())
    return dataframe

def listwise_deletion(dataframe):
    dataframe = dataframe.dropna()
    return dataframe

def imputation_median(frame, column_name=None):
    dataframe = frame
    if column_name==None:
        for i in range(11):
            dataframe = dataframe.fillna(dataframe.median().iloc[i])
    else:
        dataframe[[column_name]] = dataframe[[column_name]].fillna(dataframe[[column_name]].median())
    return dataframe

def listwise_deletion_column(dataframe, column):
    dataframe = dataframe.dropna(subset=column)
    return dataframe



## Age

#### Para el caso de esta columna se va a utilizar el metodo de imputación basado en el promedio para tomar el rango promedio de edad del data set y no implicar distinciones con sesgos subjetivos como el valor más repetido

In [18]:
imputation_mode(Data_titanic, 'Age').head()

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    NaN   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    NaN   PP 9549  16.7000    G6        S  
4    0.0    113783  26.5500  C103        S

## SibSp, Parch y Embarked

#### En el caso de estas columnas se va a utilizar la metodología de descartar los valores faltantes ya que son menores que en las demás

In [19]:
pairwise_deletion(Data_titanic, ['SibSp']).head()


PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
5           22         1       2   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
5                              Beesley, Mr. Lawrence    male  34.0    0.0   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    NaN   PP 9549  16.7000    G6        S  
5    0.0    248698  13.0000   D56        S

In [20]:
pairwise_deletion(Data_titanic, ['Parch']).head()


PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
4           12         1       1   
5           22         1       2   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    NaN   
5                              Beesley, Mr. Lawrence    male  34.0    0.0   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
4    0.0    113783  26.5500  C103        S  
5    0.0    248698  13.0000   D56        S

In [21]:
pairwise_deletion(Data_titanic, ['Embarked']).head()


PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    NaN   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    NaN   PP 9549  16.7000    G6        S  
4    0.0    113783  26.5500  C103        S

## Fare

#### Para esta columna se eliminan como primer punto los valores exagerados como primer punto para imputar el promedio en segundo lugar

In [22]:
fare = imputation_mean(Data_titanic, 'Fare')
fare = outliers_standard(Data_titanic, 'Fare', 4)
fare


PassengerId  Survived  Pclass  \
0              2         1       1   
1              4         1       1   
2              7         0       1   
3             11         1       3   
4             12         1       1   
5             22         1       2   
6             24         1       1   
7             28         0       1   
8             53         1       1   
9             55         0       1   
10            63         0       1   
11            67         1       2   
12            76         0       3   
13            89         1       1   
14            93         0       1   
15            97         0       1   
16            98         1       1   
17           103         0       1   
18           111         0       1   
19           119         0       1   
20           124         1       2   
21           125         0       1   
22           137         1       1   
23           138         0       1   
24           140         0       1   
25           149         0       2   
26           152         1       1   
27           171         0       1   
28           175         0       1   
29           178         0       1   
..           ...       ...     ...   
152          731         1       1   
154          742         0       1   
155          743         1       1   
156          746         0       1   
157          749         0       1   
158          752         1       3   
159          760         1       1   
160          764         1       1   
161          766         1       1   
162          773         0       2   
163          780         1       1   
164          782         1       1   
165          783         0       1   
166          790         0       1   
167          797         1       1   
168          803         1       1   
169          807         0       1   
170          810         1       1   
171          821         1       1   
172          824         1       3   
173          836         1       1   
174          854         1       1   
175          858         1       1   
176          863         1       1   
177          868         0       1   
178          872         1       1   
179          873         0       1   
180          880         1       1   
181          888         1       1   
182          890         1       1   

                                                  Name     Sex   Age  SibSp  \
0    Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                              McCarthy, Mr. Timothy J    male  54.0    0.0   
3                      Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                             Bonnell, Miss. Elizabeth  female  58.0    NaN   
5                                Beesley, Mr. Lawrence    male  34.0    0.0   
6                         Sloper, Mr. William Thompson       ?  24.0    0.0   
7                       Fortune, Mr. Charles Alexander       ?  19.0    3.0   
8             Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0    1.0   
9                       Ostby, Mr. Engelhart Cornelius    male  65.0    0.0   
10                         Harris, Mr. Henry Birkhardt    male  45.0    1.0   
11                        Nye, Mrs. (Elizabeth Ramell)  female  29.0    0.0   
12                             Moen, Mr. Sigurd Hansen    male  25.0    0.0   
13                          Fortune, Miss. Mabel Helen  female  24.0    3.0   
14                         Chaffee, Mr. Herbert Fuller       ?  46.0    NaN   
15                           Goldschmidt, Mr. George B    male  71.0    0.0   
16                     Greenfield, Mr. William Bertram    male  23.0    0.0   
17                           White, Mr. Richard Frasar    male  21.0    0.0   
18                      Porter, Mr. Walter Chamberlain       ?  47.0    0.0   
19                            Baxter, Mr. Quigg Edmond    male  24.0

## Limpieza a nivel Filas

In [23]:
rows = listwise_deletion(Data_titanic)
total_filas = len(rows.index)
print(total_filas)

158


#### Con la metodología Listwise se muestra que 158 filas no contienen valores faltantes

# Aplicación de Metodos individualmente

In [26]:
## Mode Imputation 

mode = Data_titanic
for i in missing:
    mode = imputation_mode(mode, i)
mode.head()

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    0.0   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    0.0   PP 9549  16.7000    G6        S  
4    0.0    113783  26.5500  C103        S

In [28]:
## Mean Imputation
mean = Data_titanic
for j in missing:
    mean = imputation_mean(mean, j)
mean.head()

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    0.0   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    0.0   PP 9549  16.7000    G6        S  
4    0.0    113783  26.5500  C103        S

In [24]:
## Listwise Deletion 
print("Cantidad de filas previo a limpieza:", len(Data_titanic.index))
print("Cantidad de filas limpias:",len(listwise_deletion(Data_titanic).index))

Cantidad de filas previo a limpieza: 183
Cantidad de filas limpias: 158


In [25]:
## Pairwise Deletion 
print("Cantidad de filas previo a limpieza:", len(Data_titanic.index))
print("Cantidad de filas limpias:",len(pairwise_deletion(Data_titanic, ['Age']).index))

Cantidad de filas previo a limpieza: 183
Cantidad de filas limpias: 183


In [30]:
## Median Imputation 

median = Data_titanic
for h in missing:
    median = imputation_median(median, h)
median.head()

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  38.0    1.0   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
2                            McCarthy, Mr. Timothy J    male  54.0    0.0   
3                    Sandstrom, Miss. Marguerite Rut  female  24.0    1.0   
4                           Bonnell, Miss. Elizabeth  female  58.0    0.0   

   Parch    Ticket     Fare Cabin Embarked  
0    0.0  PC 17599  71.2833   C85        C  
1    0.0    113803  53.1000  C123        S  
2    0.0     17463  51.8625   E46        S  
3    0.0   PP 9549  16.7000    G6        S  
4    0.0    113783  26.5500  C103        S

### La metodología que reduce los sesgos implicitos en la data de mejor manera será la eliminación, sin embargo la moda produce la perdida de la menor cantidad de data sustituyendo con los valores más populares los faltantes

#### Los metodos de imputación resultan ser más comprensibles. 

#### Se introduce sesgo al reemplazar por otros subyacentes pero la reducción de 183 a 158 resulta ser de 13.66% del total de observaciones.

## Parte 2 


#### Pre-procesamiento de la data con Missing Values y normalización de las columnas numéricas


In [31]:
df_subset = mode.drop(["Survived", 'Pclass', 'Name', 'Sex', 'Ticket', 'Cabin', 'Embarked', 'PassengerId'], axis=1)
df_normalizado=(df_subset-df_subset.mean())/df_subset.std()
for x in mode.columns:
    try:
        mode[[x]] = df_normalizado[[x]]
    except:
        pass
mode.head()

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            7         0       1   
3           11         1       3   
4           12         1       1   

                                                Name     Sex       Age  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...       ?  0.259031   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  0.060022   
2                            McCarthy, Mr. Timothy J    male  1.320412   
3                    Sandstrom, Miss. Marguerite Rut  female -0.669678   
4                           Bonnell, Miss. Elizabeth  female  1.585757   

      SibSp     Parch    Ticket      Fare Cabin Embarked  
0  0.849235 -0.585628  PC 17599 -0.101918   C85        C  
1  0.849235 -0.585628    113803 -0.343350  C123        S  
2 -0.704865 -0.585628     17463 -0.359781   E46        S  
3  0.849235 -0.585628   PP 9549 -0.826657    G6        S  
4 -0.704865 -0.585628    113783 -0.695872  C103        S

### Los valores obtenidos por la moda normalizada en las columnas numericas es similar a la imputación de la moda en los metodos anteriores